### DISCUSSION TITLE
How to add typescript on ' :rules ' prop of q-input

### DISCUSSION THREAD

**Discussion by Ethan0622** [quasarframework/quasar#10758] [General - Components / Directives / etc] [answered]
Hello, I have some trouble with q-input. I have a login component, and it include two q-input for number and password. It need to validate the user's input, so I use the ':rules" prop. 

![image](https://user-images.githubusercontent.com/67997980/133958273-243fe840-78e1-46e6-a944-01e96b9ccb3c.png)

Because my project use typescript, it has the warning " The parameter "val" implicitly has type "any". " in ' :rules ' prop. I try to add type (string) for 'val' , but it will cause a bug
![image](https://user-images.githubusercontent.com/67997980/133958714-9de7d4ff-8b28-4de1-bd1a-3c1ba3f0c499.png)
![image](https://user-images.githubusercontent.com/67997980/133958732-f8a18dbe-bd62-4fb3-9470-262dcaf2b97f.png)

so how can I add type into ':rules' and not cause the bug.  This troubles me a lot !

**Platform (please complete the following information):**
Quasar Version: v2.0.4

Quasar mode:
  - [ * ] SPA

Tested on:
  - [ * ] SPA

OS: windows10
Node: v14.15.1
NPM: 6.14.8
Browsers: Chrome 93.0.4577.82

**Comment by pdanpdan** [comment, marked as answer]
Try to define the functions in code and use the function names in code.
Even better, expose the array of rules.

**Reply by Ethan0622** [reply]
Thanks, it's a good solution! But, it may create more extra functions in code, so is it possible to write the code still in template. Or which one cause this problem, the vue or quasar ?
